openstack_openapi/openapi.rst

```diff
@@ -0,0 +1,465 @@
+based on the body.
+
+.. code:: console
+
+   $> curl -i $publicURL/servers/fake_id/action -X POST -d '
+       {
+           "lock": {"locked_reason": "I don't want to work"}
+       }
+   '
+
+RPC style actions
+-----------------
+
+.. image:: server-actions.png
+
+. . .
+
+.. danger::
+
+   OpenAPI does not support different operations based on the
+   request body
+
+So now what?
+------------
+
+“X-” extensions
+---------------
+
+OpenAPI allows *X-custom* extensions
+
+.. code:: yaml
+
+   paths:
+     /servers:
+       put:
+         ...
+         x-openstack: 
+           foo: bar
+
+
+Dynamic headers
+---------------
+
+Custom OpenAPI parameter serialization based on regex (similar to parameter serialization)
+
+.. code:: yaml
+
+   paths:
+     /foo:
+       put:
+         ...
+         parameters:
+           - in: header
+             name: X-Account-Meta-*
+             schema:
+               type: string
+             x-openstack:
+               style: regex
+
+Microversions
+-------------
+
+JSON schema ``oneOf`` with ``x-openstack`` extension and custom discriminator
+
+.. code:: yaml
+
+   components:
+     schemas:
+       VolumesCreateRequest:
+         oneOf:
+           - $ref: '#/components/schemas/VolumesCreate_30'
+           - $ref: '#/components/schemas/VolumesCreate_313'
+           - $ref: '#/components/schemas/VolumesCreate_347'
+           - $ref: '#/components/schemas/VolumesCreate_353'
+         x-openstack:
+           discriminator: microversion
+       VolumesCreate_313:
+         ...
+         x-openstack:
+           min-ver: 3.13
+           max-ver: 3.47
+       ...
+
+RPC style Actions
+-----------------
+
+JSON schema ``oneOf`` with ``x-openstack`` extension and custom discriminator
+
+.. code:: yaml
+
+   components:
+     schemas:
+       server_actions:
+         oneOf:
+           - $ref: #/components/schemas/action_foo
+           - $ref: #/components/schemas/action_bar
+         x-openstack:
+           discriminator: action
+       action_foo:
+         ...
+         x-openstack:
+           action-name: foo
+       action_bar:
+         ...
+         x-openstack:
+           action-name: bar
+
+Actions + microversions
+-----------------------
+
+Just a combination of above methods.
+
+.. code:: yaml
+
+   components:
+     schemas:
+       server_actions:
+         oneOf:
+           - $ref: #/components/schemas/action_foo
+           - $ref: #/components/schemas/action_bar
+         x-openstack:
+           discriminator: action
+       action_foo:
+         oneOf:
+           - $ref: #/components/schemas/action_foo_21
+           - $ref: #/components/schemas/action_foo_22
+         x-openstack:
+           action-name: foo
+           discriminator: microversion
+
+Constraints
+-----------
+
+- Strict body definition for certain microversion (no merging or individual
+  parameters)
+
+- Query parameters may be combined for operation marking microversion
```

Comment:
what about situation where multiple parameters (query or body) can mutually exclussive or in opposite, setting one parameter requires also another parameter to be set mandatory.
is openapi supporting definition of such situations?

Reply:
yes, as shown with dependentRequire and complex if-then-else